### PR TITLE
refactor: DfUdfAdapter to bridge ScalaUdf

### LINKED ISSUE #3665: Rewrite the UDF implementation under the new DataFusion
### What type of enhancement is this?

Other

### What does the enhancement do?

The new DataFusion deprecates how its UDF can be written before, specifically the `ScalarUDF::new`:

```rust
impl ScalarUDF {
    /// Create a new ScalarUDF from low level details.
    ///
    /// See  [`ScalarUDFImpl`] for a more convenient way to create a
    /// `ScalarUDF` using trait objects
    #[deprecated(since = "34.0.0", note = "please implement ScalarUDFImpl instead")]
    pub fn new(
        name: &str,
        signature: &Signature,
        return_type: &ReturnTypeFunction,
        fun: &ScalarFunctionImplementation,
    ) -> Self {
```

We should use the new implementation once https://github.com/GreptimeTeam/greptimedb/pull/3661 is merged.

### Implementation challenges

_No response_

### LINKED ISSUE #3666: Rewrite the UDAF implementation under the new DataFusion
### What type of enhancement is this?

Other

### What does the enhancement do?

The new DataFusion deprecates how its UDAF can be written before, specifically the `AggregateUDF::new`:

```rust
impl AggregateUDF {
    /// Create a new AggregateUDF
    ///
    /// See  [`AggregateUDFImpl`] for a more convenient way to create a
    /// `AggregateUDF` using trait objects
    #[deprecated(since = "34.0.0", note = "please implement AggregateUDFImpl instead")]
    pub fn new(
        name: &str,
        signature: &Signature,
        return_type: &ReturnTypeFunction,
        accumulator: &AccumulatorFactoryFunction,
        state_type: &StateTypeFunction,
    ) -> Self {
```

We should use the new implementation once https://github.com/GreptimeTeam/greptimedb/pull/3661 is merged.

### Implementation challenges

_No response_

### PR DESCRIPTION
**Pull request by tisonkun**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

This closes #3666 and closes #3665.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: src/common/query/src/logical_plan/udf.rs
Comment (tisonkun) [inline src/common/query/src/logical_plan/udf.rs:1]: I'd like to hear if we have other better way to implement than just lifting the adapter one level up.

But given that we define our own types, it seems the adapter solution is the most smooth way?

**Review by killme2008 — APPROVED** [review]
LGTM

File: src/common/query/src/logical_plan/udf.rs [hunk 1]
```diff
@@ -68,25 +71,61 @@ impl ScalarUdf {
...
     fn from(udf: ScalarUdf) -> Self {
-        // TODO(LFC): remove deprecated
-        #[allow(deprecated)]
-        DfScalarUDF::new(
-            &udf.name,
-            &udf.signature.into(),
-            &to_df_return_type(udf.return_type),
-            &to_df_scalar_func(udf.fun),
-        )
+        DfScalarUDF::new_from_impl(DfUdfAdapter {
```
Comment (MichaelScofield) [inline src/common/query/src/logical_plan/udf.rs:116]: There're other places that can be rewrite like this, all marked with "#[allow(deprecated)]". Can you change them as well?

File: src/common/query/src/logical_plan/udf.rs [hunk 2]
```diff
@@ -68,25 +71,61 @@ impl ScalarUdf {
...
+    return_type: datafusion_expr::ReturnTypeFunction,
+    fun: DfScalarFunctionImplementation,
+}
+
+impl Debug for DfUdfAdapter {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        f.debug_struct("DfUdfAdapter")
+            .field("name", &self.name)
+            .field("signature", &self.signature)
+            .field("fun", &"<FUNC>")
```
Comment (waynexia) [inline src/common/query/src/logical_plan/udf.rs:87]: sugg: keep it aligned with `DfUdafAdapter`?

```suggestion
            .field("fun", &"<FUNC>")
```

**Review by waynexia — APPROVED** [review]
LGTM, thanks

File: src/common/query/src/logical_plan/udf.rs (hunk 1 above)
Comment (tisonkun) [reply src/common/query/src/logical_plan/udf.rs:115]: Updated.

File: src/common/query/src/logical_plan/udf.rs (hunk 2 above)
Comment (tisonkun) [reply src/common/query/src/logical_plan/udf.rs]: Updated.

**Comment by tisonkun** [comment]
There are still some other `#[allow(deprecated)]` to be updated. I'm handling ...

**Comment by tisonkun** [comment]
All updated.

**Pull request merged**